商品購入機能

Tsuchinoko9 opened the pull request:
# What
商品購入機能を作成
・合わせてクレジットカード決済機能を作成

# Why
商品購入機能を実装するため

# 確認動画及び画像
・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/06431cd3cab1cc737a18a9a16d2d8d50

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、
　購入ページでエラーメッセージが表示される動画
https://gyazo.com/5645e0ff5544edf9a731f98d77f83bb6

・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の
　商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/84e5356f08e6dba1bdf68edf31733ce6

・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに
　遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/86729deeb8eb20e3e221899a777505bf

・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、
　商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/35d51819bbda1d5ad95d3992b516cba0

・トップページで、売却済みの商品は、画像上に「sold out」の文字が表示される動画
（※商品一覧機能実装時に未実装であったため掲載）
https://gyazo.com/523a09e9d2ceda028bf36e51567ff95c

・商品詳細ページで、売却済みの商品は、画像上に「sold out」の文字が表示される動画
（ログイン状態で、売却済み商品の商品詳細ページへ遷移した動画）
（※商品詳細機能実装時に未実装であったため掲載）
https://gyazo.com/f99d648224b81534153b84e9ff00210e

・ログイン状態の場合でも、売却済みの商品には、
　「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画
（※商品詳細機能実装時に未実装であったため掲載）
https://gyazo.com/8fcdedff74dabf8373e14f5c9f119dac

・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の
　商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
（※商品情報編集機能実装時に未実装であったため掲載）
https://gyazo.com/351dd18096bc6ef8b1fbac17a32b9fed

・テスト結果の画像
https://gyazo.com/c144c4a463b324aa8a3ce90c9caa0e39

Tsuchinoko9 committed:
「購入画面に進む」のボタンを実装

・purchase_recordsのルーティングを設定（itemsを親とするネスト記述）
・purchase_recordsのコントローラーを作成
・ordersフォルダの商品購入のビューファイルをpurchase_recordsフォルダに移動
　(用意されていたビューファイルが、購入履歴テーブルの名前が「orders」である前提のフォルダ名に入っていたため)

Tsuchinoko9 committed:
Formオブジェクトのファイルを新規作成

1つのフォームから購入情報と配送先情報をそれぞれ別テーブル(purchase_recordsテーブルとshipping_addressesテーブル)へ保存するため

Tsuchinoko9 committed:
FormオブジェクトのファイルにActiveModel::Modelを呼び出すよう記述

ActiveModel::Modelを呼び出して、form_withメソッドに対応する機能とバリデーションを行う機能を持たせる

Tsuchinoko9 committed:
Formオブジェクトのファイルにattr_accessorを記述

attr_accessorで保存したい複数のテーブルのカラム名は後で設定するため、仮のカラム名を記述

Tsuchinoko9 committed:
保存したい複数のテーブルのカラム名と各カラムに対するバリデーションを設定

(備考)このクラスにアソシエーションを定義することはできないため、belongs_toによってitem_id、user_id、purchase_record_idに関してデータが空ではないかというバリデーションはできない。そのためその旨のバリデーションを設定する必要がある。

Tsuchinoko9 committed:
各テーブルにデータを保存する処理を設定

Tsuchinoko9 committed:
shipping_addressのモデルを作成

Tsuchinoko9 committed:
各モデルにアソシエーションを記述

Tsuchinoko9 committed:
purchase_recordsのcreateアクションのルーティングを設定

Tsuchinoko9 committed:
purchase_recordsのマイグレーションファイルを編集

Tsuchinoko9 committed:
shipping_addressesのマイグレーションファイルを編集

Tsuchinoko9 committed:
shipping_addressesのデータベースを作成

・「:building」の後に余計なカンマがあってマイグレーションが実行できなかったため、カンマを削除

Tsuchinoko9 committed:
shipping_addressesのバリデーションを修正

phone_numberのカラムの型がinteger型ではなくstring型のため

Tsuchinoko9 committed:
コントローラーにindexとcreateのアクションを定義

Tsuchinoko9 committed:
purchase_record_shipping_addressのモデルファイルを修正

余計なdoを削除

Tsuchinoko9 committed:
purchase_recordsテーブルとshipping_addressesテーブルに入力フォームから情報が保存できるように設定

・Formオブジェクトを使用
・合わせてコントローラーファイル内の「purchase_record_shipping_addresses」表記を「purchase_record_shipping_address」に修正。

Tsuchinoko9 committed:
purchase_record_shipping_addressのテストファイルを作成

Tsuchinoko9 committed:
商品購入における配送先入力の単体テストを実装

・postal_codeの入力が「3桁ハイフン4桁」の半角文字列ではない値だった場合のエラーメッセージを変更（変更していたエラーメッセージを取消）
・FactoryBotはPurchaseRecordShippingAddressが(通常のモデルと違いActiveModelであるためか)save!メソッドを定義できていないとエラーが出て使用できず

Tsuchinoko9 committed:
ブラウザでアプリを開いた時にpayjp.jsが読み込まれるように設定

・payjp.jsとはクライアントサイドからPAY.JPのAPIを使用するために読み込むJavaScriptのライブラリ。この中に、安全にクレジットカード情報をトークン化するための処理などが含まれている。

Tsuchinoko9 committed:
トークン化を行うファイルを作成及びそのファイルが読み込まれるように設定

・クレジットカード情報のトークン化を行うために作成したファイルの名前は「card.js」

Tsuchinoko9 committed:
フォームの「購入」ボタンを押した時にイベントが発火するよう設定

Tsuchinoko9 committed:
クレジットカード決済機能を作成

・PAY.JPのオープンAPIを利用
・「payjp」と「gon」のGemをインストール
・商品詳細の「購入画面に進む」ボタンに「data:{turbo: false}」のオプションを指定(Javascriptの動作に影響を及ぼさない様にするため)
(備考)環境変数を設定したら「その設定したターミナル」でサーバーの再起動が必要

Tsuchinoko9 committed:
ログアウト状態で商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移するよう設定

Tsuchinoko9 committed:
ログイン状態で自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移するよう設定

Tsuchinoko9 committed:
ログイン状態で自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移するよう設定

・合わせてコントローラー内の記述をいくつか簡略化のため変更

Tsuchinoko9 committed:
商品一覧画面で、売却済みの商品は画像上に「sold out」の文字が表示されるよう設定

Tsuchinoko9 committed:
商品詳細画面で、売却済みの商品は画像上に「sold out」の文字が表示されるよう設定

Tsuchinoko9 committed:
商品詳細画面で、ログイン状態でも売却済みの商品には「商品の編集」「削除」「購入画面に進む」ボタンが表示されないように設定

Tsuchinoko9 committed:
ログイン状態でも、自身が出品した売却済み商品の商品情報編集ページへ遷移しようとするとトップページに遷移するよう設定

Tsuchinoko9 committed:
Rubocop実行により一部修正

AsukaToyoshima commented on app/controllers/purchase_records_controller.rb:
```diff
@@ -0,0 +1,65 @@
+  def purchase_record_params
+    # @item.idの値はcreateアクション前にset_item_idが実行されて取得される
+    params.require(:purchase_record_shipping_address)
+          .permit(:postal_code, :prefecture_id, :city, :addresses, :building, :phone_number, :purchase_record_id)
```
purchase_record_idは保存時に生成される値であるため、permitには不要です。削除しましょう。

AsukaToyoshima commented on app/models/purchase_record_shipping_address.rb:
```diff
@@ -0,0 +1,34 @@
+class PurchaseRecordShippingAddress
+  include ActiveModel::Model
+  attr_accessor :item_id, :user_id,
+                :postal_code, :prefecture_id, :prefecture_id, :city, :addresses, :building, :phone_number,
```
prefecture_idが重複しています。片方削除しましょう。

AsukaToyoshima commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,81 @@
+                                          postal_code: '123-4567', prefecture_id: 2, 
+                                          city: 'A市', addresses: 'B区8-9', 
+                                          building: 'Cビル10号室', phone_number: '09012345678', 
+                                          user_id: user.id, item_id: 3333)
```
PurchaseRecordShippingAddressはFactoryBotを使用して生成するようにしましょう。
またitem_idはuserと同様、FactoruyBotの値をcreateで保存してから、設定するようにしましょう。

AsukaToyoshima commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @purchase_record_shipping_address.phone_number = '090-1234-5678'
+        @purchase_record_shipping_address.valid?
+        expect(@purchase_record_shipping_address.errors.full_messages).to include("Phone number is invalid")
+      end
```
今回の要件、設定したバリデーションに合わせ、以下のテストを追加しましょう。
```
 - 電話番号が9桁以下では購入できない
 - 電話番号が12桁以上では購入できない
 - userが紐付いていなければ購入できない
 - itemが紐付いていなければ購入できない
```

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
コメントさせていただきましたので、ご確認の上、修正をお願いいたします。
疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

Tsuchinoko9 committed:
余分なコードを削除

(※コードレビュー修正)

Tsuchinoko9 committed:
異常系テストにテスト項目を追加

・それに合わせて正常系テストの文言と
「phone_numberが「10桁以上11桁以内」の半角数値でなければ商品購入はできない」のテスト時phone_numberに代入する値を変更
(※コードレビュー修正)

Tsuchinoko9 committed:
テスト用のデータ作成のためにFactoryBotを使用

・異常系テストでのitemの作成時にMySQLのエラーが出たため、
「config/environments/test.rb」にconfig.active_job.queue_adapter = :inlineを追記
(※コードレビュー修正)

Tsuchinoko9 commented:
コードレビュー、ありがとうございます！
修正致しましたのでご確認をお願い致します。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
